### ISSUE TITLE
[Feature] Ability to hold down key on Android keyboard

### ISSUE THREAD

**Issue by Slapbox** [microsoft/playwright#26796]
Specifically I'm requesting this because Android backspace behavior inside of a webview can be rather strange, especially when the user holds the backspace button down. It would be great to have a way to test this. Unfortunately I don't believe repeatedly `press`ing the button will yield similar results.

**Comment by Taythugga55** [comment]
Tayvon McFadden

**Comment by Slapbox** [comment]
Thanks for your reply @aslushnikov. Android has `beforeinput` events which don't seem to be triggered via `page.keyboard.down("Backspace")` and those are what we need to test. Android itself doesn't have any concept of backspace in the desktop-OS sense.

So while `page.keyboard.down("Backspace")` does indeed remove a character, but probably because Android Emulator provides a PC-like backspace patch for convenience. Neither `page.down` nor Backspace from PC into emulator actually produces a `beforeinput` and without the `beforeinput` event we can't run a proper test for our use case.

Specifically the code we use to trigger a single press is: `await device.input.press('Del')`as per the valid [`AndroidKeys`](https://github.com/microsoft/playwright/blob/00e65407995c2705598410ec5c79189ab5ad94e9/utils/generate_types/overrides.d.ts#L292)

**Comment by aslushnikov** [comment]
> Android has beforeinput events which don't seem to be triggered via page.keyboard.down("Backspace")

@Slapbox so I just tried, and the following works fine for me, i.e. the `beforeinput` is actually triggered for me with the following snippet:

```ts
const { _android: android } = require('playwright');

(async () => {
  const [device] = await android.devices();
  const context = await device.launchBrowser();
  const [page] = context.pages()

  // create a input element and log beforeinput event
  await page.setContent(`
    <input type="text" />
    <script>
      document.querySelector('input').addEventListener('beforeinput', e => console.log('event', e.data));
    </script>
  `);
  page.on('console', msg => console.log(msg.text()));
  await page.fill('input', 'hello');
  await page.keyboard.down('Backspace');

  await device.close()
})();
```

**Comment by Slapbox** [comment]
Thanks for testing @aslushnikov. It seems I was mistaken about exactly what the problem was; however unfortunately the `page.keyboard.down('Backspace')` still yields different results from using the software keyboard's backspace button. 

Unfortunately I'm only seeing a single character get deleted with that approach, which won't be suitable for our use case.

Additionally, although I'm less certain of this, it seems that something about what Android is doing under the hood to position the caret isn't replicated by `page.keyboard.down('Backspace')`, and so our code doesn't work as expected (ie as it would work on a real device or an emulator using the software keyboard) and it instead acts like sending a PC-based hardware Backspace into the emulator once and only once.

Is there anything else you could suggest as a possible workaround? Otherwise I think we really do, unfortunately, need such a feature.

Thanks so much again for taking the time to look into this @aslushnikov.

**Comment by aslushnikov** [comment]
> Additionally, although I'm less certain of this, it seems that something about what Android is doing under the hood to position the caret isn't replicated by page.keyboard.down('Backspace'), and so our code doesn't work as expected (ie as it would work on a real device or an emulator using the software keyboard) and it instead acts like sending a PC-based hardware Backspace into the emulator once and only once.

@Slapbox It could be that there's some other difference that is not related to Playwright automation per se. To validate or reject this claim, we would need to have a clear isolated repro for this, similarly to [this one](https://github.com/microsoft/playwright/issues/26796#issuecomment-1701485916): like, what exactly is not working? Which event is not fired?

**Comment by Slapbox** [comment]
Thanks again for your replies @aslushnikov. I recently discovered there's at least one case where `device.input.press('Del')` doesn't do what I'd expect, even though pressing it myself does, so I'll have to investigate further into what's really going on to cause that, and it may lead me to some clues about why `page.keyboard.down` doesn't work as expected for me.

**Comment by aslushnikov** [comment]
@Slapbox I'll close this for now then. Feel free to file a new issue once you have a good repro that demonstrates what doesn't work!

**Comment by Slapbox** [comment]
@aslushnikov I know this is not of much value without a repro (which I still do not have), but I am 100% sure that Playwright sends the equivalent of PC spacebar and PC backspace in at least some cases instead of sending softkeys spacebar and softkeys backspace.

I've had to make about a half-dozen exceptions in our tests to skip Android because of this and instead leave comments explaining that it's this bug.

I'm mentioning this in case anyone encounters an inkling that there might be some difference in behavior from what is expected, because I am certain there is.

For spacebar, the only difference I can find it that PC keyboard spacebar includes `evt.key === ' '` and softkeys it's `unidentified`. For backspace I can't even find a difference, but there surely must be one - I can replicate it under certain conditions. The tests fail and the tests produce the exact output of running them by hand with the PC keyboard. When I run them with the softkeys keyboard I get the expected result (which Playwright does not produce for all tests, but does for most tests)

There must be some particularities of Android keyboard which are not fully replicated by Playwright. I wish I had more to report.

**Comment by Slapbox** [comment]
@aslushnikov - I'm not certain this is the _only_ difference - but I've discovered that Playwright `keydown` and `keyup` send an `event.code === 'Backspace'` on Android, but the softkeys keyboard does not send that code.

Additionally, for `Space` - it sends a `keycode` of 32 and a `code` of `Space` - but the softkeys produce a `keycode` of 229 and no `code`

I strongly suspect these discrepancies are what's breaking many of our tests.

Do you think these findings alone would be sufficient to get this fixed? I'd love to avoid putting together a repro, but if it's necessary I certainly can.

**Comment by Slapbox** [comment]
Could anyone on the team help get this re-opened? Perhaps under another, more accurate, title?

Sorry, I don't mean to be impatient - it's just we've struggled with this limitation for many months now. Let me know if I can do anything else to help.

**Comment by Slapbox** [comment]
Friendly bump. Is there sufficient data to identify the cause of the problem? Is this mismatch in behavior considered a problem?

**Comment by Slapbox** [comment]
Should I open a new issue?

**Comment by Slapbox** [comment]
I guess I'll open a new issue unless I hear back soon. @aslushnikov

**Comment by Slapbox** [comment]
I opened #28748 for the issues I ended up finding.
